File: Algorithm/Python/409. Longest Palindrome.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-10-03 20:02:50
# @Last modified by:   WuLC
# @Last Modified time: 2016-10-03 20:03:09
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        count = {}
        for char in s:
            count.setdefault(char, 0)
            count[char] += 1
        result, odd = 0, False
        for length in count.values():
            if length % 2 == 0:
                result += length
            else:
                result += (length - 1)
                odd = True
        return result+1 if odd else result
    
        
                
            

File: Algorithm/C++/409. Longest Palindrome.cpp
/*
 * Created on Tue May 08 2018 22:17:59
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


// simple solution, not necessary to use all characters
class Solution 
{
    public:
        int longestPalindrome(string s)
        {
            unordered_map<char, int> counter;
            for(char c : s)
            {
                if (counter.count(c) > 0) counter[c]++;
                else counter[c] = 1;
            }
            int result = 0;
            bool odd = false;
            for(auto it=counter.begin(); it!=counter.end(); it++)
            {
                if((it->second & 1)) 
                {
                    result += it->second - 1;
                    odd = true;
                }
                else result += it->second;
            }
            return odd ? result+1 : result;
        }
};